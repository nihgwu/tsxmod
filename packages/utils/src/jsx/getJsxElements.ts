import {
  JsxOpeningElement,
  JsxSelfClosingElement,
  Node,
  SyntaxKind,
} from 'ts-morph'

/** Get the first descendant JsxElement based on the identifier. */
export function getJsxElement(node: Node, name: string) {
  return getJsxElements(node).find(
    (node) =>
      node.getFirstDescendantByKindOrThrow(SyntaxKind.Identifier).getText() ===
      name
  )
}

/** Get all descendant JsxElement nodes. */
export function getJsxElements(node: Node) {
  return node
    .getDescendants()
    .filter(
      (node) => Node.isJsxElement(node) || Node.isJsxSelfClosingElement(node)
    ) as (JsxOpeningElement | JsxSelfClosingElement)[]
}
