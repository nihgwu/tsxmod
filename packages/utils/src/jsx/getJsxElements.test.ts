import { readFileSync } from 'fs'
import { Project } from 'ts-morph'
import { getJsxElement, getJsxElements } from './getJsxElements'

const dialogSource = readFileSync('fixtures/app/Dialog.tsx', 'utf8')
const appSource = readFileSync('fixtures/app/App.tsx', 'utf8')

test('traces component references to the root element', () => {
  const project = new Project({ useInMemoryFileSystem: true })
  const dialogSourceFile = project.createSourceFile('Dialog.tsx', dialogSource)
  const appSourceFile = project.createSourceFile('App.tsx', appSource)

  expect(
    getJsxElements(appSourceFile).map((node) => node.getText())
  ).toMatchSnapshot()

  expect(getJsxElement(dialogSourceFile, 'Button')?.getText()).toMatchSnapshot()
})
