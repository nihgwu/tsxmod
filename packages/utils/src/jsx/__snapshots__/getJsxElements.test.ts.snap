// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`traces component references to the root element 1`] = `
[
  "<Button onClick={() => setOpen(true)} children="Open" />",
  "<Dialog onClose={() => setOpen(false)} open={open}>
        Hello Dialog
      </Dialog>",
]
`;

exports[`traces component references to the root element 2`] = `"<Button onClick={props.onClose}>Close</Button>"`;
